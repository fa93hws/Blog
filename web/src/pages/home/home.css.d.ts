// This file is automatically generated.
// Please do not change this file!
interface CssExports {
  'blogAbstract': string;
  'blogList': string;
  'gridBaseline': string;
  'home': string;
  'monoGreyLight': string;
}
export const cssExports: CssExports;
export default cssExports;
