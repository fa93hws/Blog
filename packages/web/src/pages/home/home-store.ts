import { makeObservable, observable, action } from 'mobx';
import type { IPostService } from '@services/post/post-service';
import type { PostProto } from '@fa93hws-blog/protos';
import type { ShowFn } from '@components/alert/alert';

type Summary = PostProto.ISummary;
export class HomeStore {
  posts: readonly Summary[] = [];

  loading = false;

  constructor(
    private readonly postService: IPostService,
    private readonly showGlobalMsg: ShowFn,
  ) {
    makeObservable(this, {
      posts: observable.ref,
      loading: observable.ref,
      setPosts: action,
      setLoading: action,
    });
  }

  setPosts(postList: readonly Summary[]) {
    this.posts = postList;
  }

  setLoading(loading: boolean) {
    this.loading = loading;
  }

  async fetchList() {
    this.setLoading(true);
    const result = await this.postService.fetchList();
    this.setLoading(false);
    if (result.isOk) {
      this.setPosts(result.value.list ?? []);
    } else {
      this.setPosts([]);
      this.showGlobalMsg('error', '载入博客列表失败');
    }
  }
}
