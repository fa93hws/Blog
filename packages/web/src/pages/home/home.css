@value gridBaseline from '~components/metrics.css';
@value monoGreyLight from '~components/colors.css';

.home {
  min-height: 100vh;
  background-color: monoGreyLight;
}

.blogList {
  width: calc(gridBaseline * 120);
  margin: 0 auto;
  padding: calc(gridBaseline * 4) 0;
}

.blogAbstract {
  margin-bottom: calc(gridBaseline * 4);
}
