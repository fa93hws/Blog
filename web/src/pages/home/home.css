@value monoGreyLight from '~components/colors.css';

.home {
  height: 100vh;
  background-color: monoGreyLight;
}
