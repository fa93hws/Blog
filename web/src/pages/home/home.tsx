import * as React from 'react';
import { BlogListProto } from 'services/blog-list/blog-list.proto';
import { AbstractCard } from './abstract-card/card';
import styles from './home.css';

type HomePageProps = {
  blogList: BlogListProto;
};

// TODO Change to loading state
const emptyBlogList: BlogListProto = {
  currentPage: 1,
  totalPage: 1,
  blogAbstracts: [],
};

export const HomePage = React.memo(({ blogList }: HomePageProps) => (
  <div className={styles.home}>
    <ul className={styles.blogList}>
      {blogList.blogAbstracts.map(blogAbstract => (
        <li key={blogAbstract.id} className={styles.blogAbstract}>
          <AbstractCard
            time={blogAbstract.time}
            title={blogAbstract.title}
            abstract={blogAbstract.abstract}
            linkTo={`/blog/${blogAbstract.id}`}
          />
        </li>
      ))}
    </ul>
  </div>
));

export function createHomePage() {
  const ConnectedHomePage = () => {
    const [blogList, setBlogList] = React.useState(emptyBlogList);
    React.useEffect(() => {
      import(
        'services/blog-list/fake/fake-blog-list'
      ).then(({ fakeBlogList }) => setBlogList(fakeBlogList));
    }, [setBlogList]);
    return <HomePage blogList={blogList} />;
  };
  return ConnectedHomePage;
}
