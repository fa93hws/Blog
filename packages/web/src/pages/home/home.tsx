import * as React from 'react';
import { observer } from 'mobx-react';
import type { PostService } from '@services/post/post-service';
import type { PostInList } from '@services/post/post';
import { HomeStore } from './home-store';
import { PostList } from './post-list/post-list';

type HomeProps = {
  posts: readonly PostInList[];
  onMount(): void;
};
export const Home = React.memo((props: HomeProps) => {
  React.useEffect(() => {
    props.onMount();
  }, []);
  return <PostList posts={props.posts} />;
});

export function createHome(postService: PostService) {
  const store = new HomeStore(postService);
  const onMount = () => store.fetchList();
  return observer(() => <Home posts={store.posts} onMount={onMount} />);
}
