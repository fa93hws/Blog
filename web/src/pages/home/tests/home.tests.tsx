import * as React from 'react';
import {
  BlogListProto,
  BlogAbstractProto,
} from 'services/blog-list/blog-list.proto';
import { MemoryRouter } from 'react-router-dom';
import { mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import { createHomePage, HomePage } from '../home';

describe('HomePage', () => {
  const blogAbstracts: BlogAbstractProto[] = [
    {
      abstract: 'abstract1',
      time: '1999年2月2日',
      title: 'title1',
      id: 'guming1',
    },
    {
      abstract: 'abstract2',
      time: '1999年2月2日',
      title: 'title2',
      id: 'guming2',
    },
  ];

  it('renders the homepage at page 1', () => {
    const list: BlogListProto = {
      currentPage: 1,
      totalPage: 10,
      blogAbstracts,
    };
    expect(
      <MemoryRouter>
        <HomePage blogList={list} />
      </MemoryRouter>,
    ).toMatchRenderedSnapshot();
  });

  it('mounts the stateful homepage without error', async () => {
    const HomePageImpl = createHomePage();
    const mountPromise = act(async () => {
      mount(
        <MemoryRouter>
          <HomePageImpl />
        </MemoryRouter>,
      );
    });
    // linting is wrong
    // eslint-disable-next-line jest/valid-expect
    expect(mountPromise).resolves.toBe(undefined);
  });
});
