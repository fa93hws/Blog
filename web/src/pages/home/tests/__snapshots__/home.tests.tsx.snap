// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage renders the homepage at page 1 1`] = `
<div
  class="home"
>
  <ul
    class="blogList"
  >
    <li
      class="blogAbstract"
    >
      <article
        class="card"
      >
        <div
          class="time"
        >
          <p
            class="textMicro"
          >
            1999年2月2日
          </p>
        </div>
        <div
          class="title"
        >
          <div
            class="title"
          >
            <a
              href="/blog/guming1"
            >
              <h3
                class="titleMedium"
              >
                title1
              </h3>
            </a>
            <div
              class="underline"
            />
          </div>
        </div>
        <div
          class="content"
        >
          <p
            class="textSmall"
          >
            abstract1
          </p>
        </div>
      </article>
    </li>
    <li
      class="blogAbstract"
    >
      <article
        class="card"
      >
        <div
          class="time"
        >
          <p
            class="textMicro"
          >
            1999年2月2日
          </p>
        </div>
        <div
          class="title"
        >
          <div
            class="title"
          >
            <a
              href="/blog/guming2"
            >
              <h3
                class="titleMedium"
              >
                title2
              </h3>
            </a>
            <div
              class="underline"
            />
          </div>
        </div>
        <div
          class="content"
        >
          <p
            class="textSmall"
          >
            abstract2
          </p>
        </div>
      </article>
    </li>
  </ul>
</div>
`;
